packages/backend/ubuntu-pro.go

```diff
@@ -0,0 +1,186 @@
+    enabled, err := isServiceEnabled("esm_2dapps")
+    return &pb.Boolean{ Ret: enabled }, err
+}
+
+func (s *ProServer) IsKernelLivePatchEnabled (ctx context.Context, _ *emptypb.Empty) (*pb.Boolean, error) {
+    enabled, err := isServiceEnabled("livepatch")
+    return &pb.Boolean{ Ret: enabled }, err
+}
+
+var dbusServices map[dbus.ObjectPath]map[string]map[string]dbus.Variant
+func connectToSystemBus() error {
+    var err error
+    conn, err = dbus.ConnectSystemBus()
+    return err
+}
+
+func enableService(basename string, able string) error {
+    /* XXX: Would this be better? Can the paths change?
+
+    nameToPath := make(map[string]dbus.ObjectPath)
+    bus := conn.Object("com.canonical.UbuntuAdvantage", "/")
+    err = bus.Call("org.freedesktop.DBus.ObjectManager.GetManagedObjects", 0).Store(&dbusServices)
+    if err != nil {
+        return err
+    }
+    for pat, properties := range dbusServices {
+        path := string(pat)
+        if strings.HasPrefix(path, "/com/canonical/UbuntuAdvantage/Services/") {
+            idx := properties["com.canonical.UbuntuAdvantage.Service"]["Name"].String()
+            nameToPath[idx] = pat
+        }
+    }
+    obj := conn.Object("com.canonical.UbuntuAdvantage", nameToPath[arag])
+
+    */
+
+    obj := conn.Object(
+        "com.canonical.UbuntuAdvantage",
+        dbus.ObjectPath("/com/canonical/UbuntuAdvantage/Services/" + basename),
+    )
+    call := obj.Call(
+        "com.canonical.UbuntuAdvantage.Service." + able,
+        dbus.FlagAllowInteractiveAuthorization,
+    )
+    if call.Err != nil {
+        return call.Err
+    }
+    return nil
+}
+
+func (s *ProServer) EnableKernelLivePatch (ctx context.Context, _ *emptypb.Empty) (*emptypb.Empty, error) {
+    return new(emptypb.Empty), enableService("livepatch", "Enable")
+}
+
+func (s *ProServer) DisableKernelLivePatch (ctx context.Context, _ *emptypb.Empty) (*emptypb.Empty, error) {
+    return new(emptypb.Empty), enableService("livepatch", "Disable")
+}
+
+func (s *ProServer) EnableEsmApps (ctx context.Context, _ *emptypb.Empty) (*emptypb.Empty, error) {
+    return new(emptypb.Empty), enableService("esm_2dapps", "Enable")
+}
+
+func (s *ProServer) DisableEsmApps (ctx context.Context, _ *emptypb.Empty) (*emptypb.Empty, error) {
+    return new(emptypb.Empty), enableService("esm_2dapps", "Disable")
+}
+
+func (s *ProServer) EnableInfra (ctx context.Context, _ *emptypb.Empty) (*emptypb.Empty, error) {
+    return new(emptypb.Empty), enableService("esm_2dinfra", "Enable")
+}
+
+func (s *ProServer) DisableInfra (ctx context.Context, _ *emptypb.Empty) (*emptypb.Empty, error) {
+    return new(emptypb.Empty), enableService("esm_2dinfra", "Disable")
+}
+
+func pollProInitiate(reqId string) {
+    cmd := exec.Command("pro", "api", "u.pro.attach.magic.wait.v1", "--args", "magic_token=" + reqId)
+    out, err := cmd.Output()
+    if err != nil {
```

Comment:
The `pro api <wait>` command might possibly fail because of 

- Failure exec'ing per se, e.g. inexistent command or insufficient permissions etc. (less likely).
- Failure of communication to the server, e.g. machine lost connection to Internet, in case it returns a number of descriptive error codes (more likely).

This first err is catching both above scenarios, defeating the usefulness of the error code, which would be caught by the next `if` check.

Any suggestion as to the best way of addressing this?

Reply:
Not something I've had to deal with to much, but if I look at [how the enterprise team is handling similar calls](https://github.com/canonical/ubuntu-pro-for-wsl/blob/main/wsl-pro-service/internal/system/pro.go#L10-L27) it looks like they are including the context and just giving a single error along with the context and allowing that to suffice for multiple error cases :slightly_smiling_face:

Reply:
I dealt with it in another way, but it is more robust now, since it also checks whether the response is indeed a JSON.